File: src/cpr_detector.py
from enum import Flag, auto


class DetectorState(Flag):
    ''''''
    EMPTY = auto()
    FIRST = auto()
    SECOND = auto()
    THIRD = auto()
    FOURTH = auto()
    FIFTH = auto()
    SIXTH = auto()
    SEVENTH = auto()
    EIGHTH = auto()
    MATCH = auto()


def CPRNumber:
    def __init__(self, cpr):
        self._cpr = cpr

    def __str__(self):
        return ""


def init_cpr():
    ''''''
    return {
            0: None, 1: None, 
            2: None, 3: None, 
            4: None, 5: None, 
            6: None, 7: None, 
            8: None, 9: None, 
            }


def digits():
    ''''''
    return "0" + nonzero_digits()

def nonzero_digits():
    ''''''
    return "123456789"


def delimiters():
    ''''''
    return ' -/'


class CPRDetector:
    ''''''
    def __init__(self):
        self._state = DetectorState.EMPTY

    def _reset(self):
        self._state = DetectorState.EMPTY

    def _update(self, c: str, new_state: DetectorState, accepted):
        if c in accepted:
            self._state = new_state
            return c
        else:
            self._reset()
            return None

    def _check_day_month(self, cpr):
        day = int(cpr[0] + cpr[1])
        month = int(cpr[2] + cpr[3])

        match month:
            case 2 | 4 | 6 | 9 | 11 if day > 30:
                self._reset()
            case 2 if day > 29:
                self._reset()
            case 2 if day == 29:
                return True

        return False

    def find_match(self, content: str, reset: bool = True):
        if content is None or len(content) < 10:
            return

        cpr = init_cpr()
        previous = None
        for i, c in enumerate(content):
            match self._state:
                case DetectorState.EMPTY:
                    previous = cpr[0] = self._update(c, DetectorState.FIRST, '0123')
                    if previous:
                        begin = i

                case DetectorState.FIRST:
                    accepted = []
                    if previous in "12":
                        accepted = digits()
                    elif previous == "0":
                        accepted = nonzero_digits()
                    elif previous == "3":
                        accepted = '0' | '1'

                    previous = cpr[1] = self._update(c, DetectorState.SECOND, accepted)

                case DetectorState.SECOND:
                    previous = cpr[2] = self._update(c, DetectorState.THIRD, '01')

                case DetectorState.THIRD:
                    accepted = []
                    match previous:
                        case '0':
                            accepted = nonzero_digits()
                        case '1':
                            accepted = '012'
                            
                    previous = cpr[3] = self._update(c, DetectorState.FOURTH, accepted)
                    leap_year = self._check_day_month(cpr)

                case DetectorState.FOURTH:
                    previous = cpr[4] = self._update(c, DetectorState.FIFTH, digits())

                case DetectorState.FIFTH:
                    accepted = []
                    match previous:
                        case '0':
                            accepted = nonzero_digits()
                        case '1':
                            accepted = digits()

                    previous = cpr[5] = self._update(c, DetectorState.SIXTH, accepted)

                    if leap_year:
                        year = int(cpr[4] + cpr[5])
                        if year % 4 != 0:
                           self._reset() 

                case DetectorState.SIXTH:
                    previous = cpr[6] = self._update(c, DetectorState.SEVENTH, accepted)
                case DetectorState.SEVENTH:
                    previous = cpr[7] = self._update(c, DetectorState.EIGHTH, accepted)
                case DetectorState.EIGHTH:
                    previous = cpr[8] = self._update(c, DetectorState.MATCH, accepted)
                case DetectorState.MATCH:
                    previous = cpr[9] = self._update(c, DetectorState.MATCH, accepted)

                    control = int(cpr[6] + cpr[7] + cpr[8] + cpr[9])

                    if control <= 0:
                        self._reset()
                        continue

                    return {
                        "cpr": cpr,
                        "begin": begin,
                        "end": i,
                        }

File: cpp/src/cpr-detector.cpp
#include <cpr-detector.hpp>


namespace CPRDetector {

  constexpr CPRDetector::CPRDetector(CPRDetector&& other) noexcept
    : state_(std::move(other.state_)), check_mod11_(std::move(other.check_mod11_)) {}

  constexpr CPRDetector& CPRDetector::operator=(CPRDetector&& other) noexcept {
    if (this != &other) {
      std::swap(other.check_mod11_, check_mod11_);
      state_ = std::move(other.state_);
    }

    return *this;
  }

  void CPRDetector::reset() noexcept {
    // Set the detector state to Empty.
    state_ = CPRDetectorState::Empty;
  }

  char CPRDetector::update(char c, CPRDetectorState new_state, Predicate is_acceptable) noexcept {
    if (is_acceptable(c)) {
      // If c is in the set of acceptable tokens, change state and return c.
      state_ = new_state;
      return c;
    } else {
      // Reset detector state and return 0.
      reset();
      return 0;
    }
  }

  bool CPRDetector::check_day_month(const std::string& cpr) noexcept {
    // Convert the first four digits representing day and month to ints.
    int day = std::stoi(std::string(cpr, 0, 2));
    int month = std::stoi(std::string(cpr, 2, 2));

    if (month == 2) {
      if (day == 29)
	// It is February 29th. Raise a flag to indicate that this should be a leap year.
	return true;
      else if (day > 29)
	// February 30th and 31st are invalid dates, so do a reset.
	reset();
    } else if (day > 30 &&
	       !((month > 7 && month % 2 == 0)
		|| (month < 8 && month % 2 != 0))) {
      // The 31st of April, June, September or November are invalid dates.
      reset();
    } 
    
    // We can't tell from the day-month combination if this should be a leap year.
    return false;
  }

  void CPRDetector::check_leap_year(const std::string& cpr) noexcept {
    // Convert the digits representing a year to an int.
    int year = std::stoi(std::string(cpr, 4, 2));

    // If it is not a leap year, then reset.
    // Note that since we cannot tell the century, we assume that 00 is a leap year.
    if (year % 4 != 0)
      reset();
  }

  void CPRDetector::check_and_append_cpr(std::string& cpr, CPRResults& results, size_t begin, size_t end) noexcept {
    // Convert the 4 control digits to an int.
    int control = std::stoi(std::string(cpr, 6, 4));

    // We reject the control sequence '0000'.
    if (control > 0) {
      CPRResult result(cpr, begin, end);
      results.push_back(result);
    }
  }

  CPRResults CPRDetector::find_matches(const std::string& content) noexcept {
    CPRResults results;

    // If the content is too short to contain a match, return early.
    if (content.size() < 10) {
      return results;
    }

    // Initialize.
    std::string cpr(10, 0);
    char previous = 0;
    size_t begin = 0;
    bool allow_separator = false;
    bool leap_year = false;
    Predicate is_acceptable = [](char) { return false; };

    for(auto it = std::begin(content); it != std::end(content); ++it) {
      switch (state_) {
      case CPRDetectorState::Empty:
	is_acceptable = make_predicate('0', '1', '2', '3');
	previous = update(*it, CPRDetectorState::First, is_acceptable);

	if (previous != 0) {
	  cpr[0] = previous;
	  begin = std::distance(content.begin(), it);
	}

	break;
      case CPRDetectorState::First:
	if (previous == '0') {
	  is_acceptable = is_nonzero_digit;
	} else if (previous == '1' || previous == '2') {
	  is_acceptable = is_digit;
	} else if (previous == '3') {
	  is_acceptable = make_predicate('0', '1');
	} else {
	  reset();
	  previous = 0;
	  continue;
	}

	previous = cpr[1] = update(*it, CPRDetectorState::Second, is_acceptable);

	if (previous != 0)
	  // Next time, we allow a space.
	  allow_separator = true;

	break;
      case CPRDetectorState::Second:
	if (is_space(*it) && allow_separator) {
	  // Skip a space character.
	  allow_separator = false;
	  continue;
	}

	is_acceptable = make_predicate('0', '1');
	previous = cpr[2] = update(*it, CPRDetectorState::Third, is_acceptable);

	break;
      case CPRDetectorState::Third:
	if (previous == '0') {
	  is_acceptable = is_nonzero_digit;
	} else if (previous == '1') {
	  is_acceptable = make_predicate('0', '1', '2');
	} else {
	  reset();
	  previous = 0;
	  continue;
	}

	previous = cpr[3] = update(*it, CPRDetectorState::Fourth, is_acceptable);

	leap_year = check_day_month(cpr);

	if (previous != 0)
	  // Next time, we allow a space.
	  allow_separator = true;

	break;
      case CPRDetectorState::Fourth:
	if (is_space(*it) && allow_separator) {
	  // Skip a space character.
	  allow_separator = false;
	  continue;
	}

	is_acceptable = is_digit;

	previous = cpr[4] = update(*it, CPRDetectorState::Fifth, is_acceptable);
	
	break;
      case CPRDetectorState::Fifth:
	if (previous == '0') {
	  is_acceptable = is_nonzero_digit;
	} else if (previous == '1') {
	  is_acceptable = is_digit;
	} else {
	  reset();
	  previous = 0;
	  continue;
	}

	previous = cpr[5] = update(*it, CPRDetectorState::Sixth, is_acceptable);

	if (leap_year) 
	  check_leap_year(cpr);
       
	if (previous != 0) 
	  // Next time we allow one of the valid separators.
	  allow_separator = true;
	
	break;
      case CPRDetectorState::Sixth:
	if (allow_separator && is_separator(*it)) {
	  // Skip one of the valid separator characters.
	  allow_separator = false;
	  continue;
	}

	is_acceptable = is_digit;
	previous = cpr[6] = update(*it, CPRDetectorState::Seventh, is_acceptable);
	
	break;
      case CPRDetectorState::Seventh:
	is_acceptable = is_digit;
	previous = cpr[7] = update(*it, CPRDetectorState::Eighth, is_acceptable);

	break;
      case CPRDetectorState::Eighth:
	is_acceptable = is_digit;
	previous = cpr[8] = update(*it, CPRDetectorState::Match, is_acceptable);

	break;
      case CPRDetectorState::Match:
	is_acceptable = is_digit;
	previous = cpr[9] = update(*it, CPRDetectorState::Match, is_acceptable);

	check_and_append_cpr(cpr, results, begin, std::distance(content.begin(), it));

	previous = 0;
	reset();

	break;
      default:
	reset();
	break;
      }
    }

    return results;
  }
};

